pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ControllerLeaderLocatorIntegrationTest.java

```diff
@@ -0,0 +1,164 @@
+    Assert.assertEquals((int) pair.getSecond(), ControllerTest.DEFAULT_CONTROLLER_PORT);
+
+    // Enable lead controller resource
+    enableResourceConfigForLeadControllerResource(true);
+
+    // Mock the behavior that 40 seconds have passed.
+    controllerLeaderLocator.setLastCacheInvalidateMillis(System.currentTimeMillis() - 40_000L);
+    controllerLeaderLocator.invalidateCachedControllerLeader();
+
+    // After resource config is enabled, use the lead controller in the resource
+    pair = controllerLeaderLocator.getControllerLeader(testTableName1);
+    Assert.assertNotNull(pair);
+    Assert.assertEquals(pair.getFirst(), ControllerTest.LOCAL_HOST);
+    Assert.assertEquals((int) pair.getSecond(), ControllerTest.DEFAULT_CONTROLLER_PORT);
+
+    // Fetch the pair for test table 2, which should be the same as the 1st pair.
+    Pair<String, Integer> secondPair = controllerLeaderLocator.getControllerLeader(testTableName2);
+    Assert.assertNotNull(secondPair);
+    Assert.assertEquals(pair.getFirst(), secondPair.getFirst());
+    Assert.assertEquals(pair.getSecond(), secondPair.getSecond());
+
+    ControllerConf secondControllerConfig = getDefaultControllerConfiguration();
+    secondControllerConfig.setControllerPort(Integer.toString(DEFAULT_CONTROLLER_PORT + 1));
+    ControllerStarter secondControllerStarter = new ControllerStarter(secondControllerConfig);
+    secondControllerStarter.start();
+
+    TestUtils
+        .waitForCondition(aVoid -> secondControllerStarter.getHelixResourceManager().getHelixZkManager().isConnected(),
+            TIMEOUT_IN_MS, "Failed to start the second controller");
+
+    // Generate a table name that the second controller is its new lead controller, which should be different from the first table name.
+    AtomicReference<String> testTableName3 = new AtomicReference<>();
+    TestUtils.waitForCondition(aVoid -> {
+      String tableName;
+      try {
+        tableName = generateTableNameUsingSecondController(secondControllerConfig, testTableName1);
+      } catch (Exception e) {
+        return false;
+      }
+      testTableName3.set(tableName);
+      return tableName != null;
+    }, TIMEOUT_IN_MS, "Failed to find the second table");
+
+    // Mock the behavior that 40 seconds have passed.
+    controllerLeaderLocator.setLastCacheInvalidateMillis(System.currentTimeMillis() - 40_000L);
+    controllerLeaderLocator.invalidateCachedControllerLeader();
+
+    // The second controller should be the lead controller for test table 3, which isn't the helix leader.
+    Pair<String, Integer> thirdPair = controllerLeaderLocator.getControllerLeader(testTableName3.get());
+    Assert.assertNotNull(thirdPair);
+    Assert.assertEquals(thirdPair.getFirst(), ControllerTest.LOCAL_HOST);
+    Assert.assertEquals((int) thirdPair.getSecond(), (ControllerTest.DEFAULT_CONTROLLER_PORT + 1));
+    Assert.assertNotEquals(pair.getSecond(), thirdPair.getSecond());
+
```

Comment:
Add another step here (in the same test) to disable the resource and check if we are getting helix leader now.

Reply:
Test added.